include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOpConversionHelper.h

```diff
@@ -0,0 +1,72 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "mlir-tcp/Dialect/IR/TcpOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+
+namespace mlir::tcp {
+
+/// Helper class for converting torch ops to tcp.custom_ops. Encapsulates a
+/// few common patterns which we see being used in these conversions:
+///
+/// 1. Some of the tensor operands to the original torch op are converted
+/// into positional tensor operands. Given that a tcp.custom_op does not
+/// have a strongly typed API to fetch the operands other than by index, we
+/// also add an extra StringArray attribute to the tcp.custom_op called
+/// "torch_operand_names" which describes what each positional argument of
+/// the `tcp.custom_op` corresponds to. This also allows us to handle the
+/// case of optional tensor arguments.
+///
+/// 2. Some of the tensor operands are converted into named attributes of
```

Comment:
I believe this should say `non-tensor operands`, since we keep all tensor operands as-is.

Reply:
I'll change "tensor operands" to just "operands". My understanding is that a torch OP _can_ have a tensor operand which we merely insist is a constant.

Reply:
>  My understanding is that a torch OP can have a tensor operand which we merely insist is a constant.

FWICT we always preserve tensor operands as tensors. Yes, scalar types get converted to attributes, sometimes even if they're not constants. @navahgar Can you think of a case otherwise?

[TorchTypes.td](https://sourcegraph.com/github.com/llvm/torch-mlir@main/-/blob/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td) has a clear separation between tensor and non-tensor (aka scalar) dtypes. The regular types like `Torch_IntType` or `Torch_FloatType` are explicitly modeled after python ints/floats and it notes `This type is not used for modeling tensor dtypes`. 
```
def Torch_IntType : Torch_Type<"Int", "int"> {
  let summary = "Torch IntType";
  let description = [{
    The integer type used to model the Python `int` type in TorchScript.
    TorchScript itself models this type as a 64-bit signed integer.

    Note: This type is not used for modeling tensor dtypes.
  }];
}
```
Please correct me if I'm missing something.

Reply:
> > My understanding is that a torch OP can have a tensor operand which we merely insist is a constant.
> 
> FWICT we always preserve tensor operands as tensors. 

That's right. Tensor operands are always inputs to the ops. At this point during the lowering, I can't think of a case where we insist this to be constants.

But later in the compilation, we have some cases, like in the QuantSim ops, where we insist that they are constants.

> Yes, scalar types get converted to attributes, sometimes even if they're not constants.

One clarification here: they have to be constants to be converted to attributes. We probably error out if they are not constants in such cases.